report.py

```diff
@@ -227,7 +227,13 @@ def run_report(config, blockers, server, header, test, save):
 			smtp.ehlo_or_helo_if_needed()
 
 			# send email to all addresses
-			smtp.sendmail(msg['From'], recipients, msg.as_string())
+			response = smtp.sendmail(msg['From'], recipients, msg.as_string())
+
+			# log success if all recipients recieved report, otherwise raise exception
+			if response == {}:
+				print("Report successfully delivered to all recipients")
+			else:
+				raise Exception("Failed to deliver report to following recipients: {}".format(response))
```

Comment:
Please see https://docs.python.org/3/library/smtplib.html#smtplib.SMTP.sendmail which states:

"This method will return normally if the mail is accepted for at least one recipient. Otherwise it will raise an exception. That is, if this method does not raise an exception, then someone should get your mail. If this method does not raise an exception, it returns a dictionary, with one entry for each recipient that was refused. Each entry contains a tuple of the SMTP error code and the accompanying error message sent by the server."

So it's good you are logging (I assume this exception gets logged) the response. I assume also that the format(response) unpacks the dictionary. I'd add this to a finally of a larger exception catch. Note the following:

SMTPRecipientsRefused
All recipients were refused. Nobody got the mail. The recipients attribute of the exception object is a dictionary with information about the refused recipients (like the one returned when at least one recipient was accepted).

SMTPHeloError
The server didn’t reply properly to the HELO greeting.

SMTPSenderRefused
The server didn’t accept the from_addr.

SMTPDataError
The server replied with an unexpected error code (other than a refusal of a recipient).

SMTPNotSupportedError
SMTPUTF8 was given in the mail_options but is not supported by the server.

So you could attempt to catch each of the the above and then finally log the dictionary which was returned.

Reply:
> So it's good you are logging (I assume this exception gets logged) the response. I assume also that the format(response) unpacks the dictionary.

Yes and yes

> I'd add this to a finally of a larger exception catch. Note the following:
> 
> SMTPRecipientsRefused
> All recipients were refused. Nobody got the mail. The recipients attribute of the exception object is a dictionary with information about the refused recipients (like the one returned when at least one recipient was accepted).
> 
> SMTPHeloError
> The server didn’t reply properly to the HELO greeting.
> 
> SMTPSenderRefused
> The server didn’t accept the from_addr.
> 
> SMTPDataError
> The server replied with an unexpected error code (other than a refusal of a recipient).
> 
> SMTPNotSupportedError
> SMTPUTF8 was given in the mail_options but is not supported by the server.
> 
> So you could attempt to catch each of the the above and then finally log the dictionary which was returned.

So as of now, all of the above exceptions are raised either here
```python
response = smtp.sendmail(msg['From'], recipients, msg.as_string())
```
or here
```python
else:
	raise Exception("Mail server cannot deliver report to following recipients: {}".format(response))

```
And caught here

```python
except Exception as e:
	print("Error sending email report: {}\nHTML file generated".format(e))
	generate_html_file(htmlcode)
```
Do you think this needs to be more explicit, i.e. different logged messages for different exceptions?

Reply:
That would make it better since you'd capture more specifically what went wrong but what you have is an improvement. If you don't want to capture the additional details, then you don't NEED to.

Reply:
Hmmm so right now the captured exception itself should be printed, although you're right it would be better if each had a uniquely handled block. Would you be amicable to starting with this patch and then adding in more detailed logging at a later date?